<a href="https://colab.research.google.com/github/bihjane/Data-science-and-AI-K-12-teaching-tools/blob/main/Copy_of_4_5_%5BResource%5D_Basic_Statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic Statistics in Python

To calculate basic descriptive statistics in Python, we are going to use ```pandas```.

Other packages, like ```statistics``` could be helpful for more advanced statistics. However, ```pandas``` can do all of the basic calculations.

In [ ]:
import pandas as pd

import gdown
url = 'https://drive.google.com/file/d/1iz_fEGAVvAlT0eE44ia_YvFTnFvTrD44/view'
output_path = 'roller_coaster.csv'
gdown.download(url, output_path, quiet=False, fuzzy=True)

Downloading...
From: https://drive.google.com/uc?id=1iz_fEGAVvAlT0eE44ia_YvFTnFvTrD44
To: /content/roller_coaster.csv
100%|██████████| 14.7k/14.7k [00:00<00:00, 6.67MB/s]


'roller_coaster.csv'

##Exploring Data

First, upload and clean the data

(Remember to upload ```roller_coasters.csv```)

(```coaster_df.dropna(inplace = True)``` is the same thing as ```coaster_df = coaster_df.dropna()```)

**Option 1**: This drops any entry that has *any* missing data.

In [ ]:
coaster_df = pd.read_csv('/content/roller_coaster.csv')
coaster_df.info()
coaster_df.dropna(inplace = True)
coaster_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Coaster            157 non-null    object 
 1   Park               157 non-null    object 
 2   Top_Speed          146 non-null    float64
 3   Max_Height         152 non-null    float64
 4   Drop               154 non-null    float64
 5   Length             154 non-null    float64
 6   Duration           129 non-null    float64
 7   Type               157 non-null    object 
 8   Design             157 non-null    object 
 9   Year_Opened        157 non-null    int64  
 10  Age_Group          157 non-null    object 
 11  Inversions         156 non-null    object 
 12  Num_of_Inversions  156 non-null    float64
 13  City               157 non-null    object 
dtypes: float64(6), int64(1), object(7)
memory usage: 17.3+ KB
<class 'pandas.core.frame.DataFrame'>
Index: 115 entries, 2 to 1

**Option 2:** This only drops items who are missing their ```Top_Speed```, which is the column I want.

In [ ]:
coaster_df = pd.read_csv('/content/roller_coaster.csv')
coaster_df.info()
coaster_df.dropna(inplace = True, subset= ['Top_Speed'])
coaster_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Coaster            157 non-null    object 
 1   Park               157 non-null    object 
 2   Top_Speed          146 non-null    float64
 3   Max_Height         152 non-null    float64
 4   Drop               154 non-null    float64
 5   Length             154 non-null    float64
 6   Duration           129 non-null    float64
 7   Type               157 non-null    object 
 8   Design             157 non-null    object 
 9   Year_Opened        157 non-null    int64  
 10  Age_Group          157 non-null    object 
 11  Inversions         156 non-null    object 
 12  Num_of_Inversions  156 non-null    float64
 13  City               157 non-null    object 
dtypes: float64(6), int64(1), object(7)
memory usage: 17.3+ KB
<class 'pandas.core.frame.DataFrame'>
Index: 146 entries, 0 to 1

**General Exploration**
Remember that we use ```info()```, ```describe()```, and ```head()``` to explore.

In [ ]:
coaster_df.info()

In [ ]:
coaster_df.describe()

In [ ]:
coaster_df.head()

**Finding the number of samples**: We can find the number of non-missing samples in one column using ```.count()```.
* We have 146 rows if we only removed missing values in ```Top_Speed```
* We have 115 rows if we dropped all missing values

In [ ]:
speeds = coaster_df['Top_Speed']

speeds.count()

**Finding the max/min**: ```max()``` and ```min()``` can find the max and min of a column

In [ ]:
max = speeds.max()
min = speeds.min()

print('maximum:', max)
print('minimum:', min)

**Getting a Subset** Recall that we used subsets when we were creating visualizations. Suppose I want to only consider older roller coasters:

In [ ]:
old_coasters = coaster_df[coaster_df["Age_Group"] == "older"]

We can recalculate to learn more about our subset:

In [ ]:
old_speeds = old_coasters['Top_Speed']

print('number of samples:', old_speeds.count())
print('min:', old_speeds.max())
print('max:',old_speeds.min())

## Calculate Statistics

Pandas has a built-in method to calculate descriptive statistics of columns. Run this code block to see how it works.

In [ ]:
coaster_df.describe()

Let's calculate the descriptive statistics of the Top Speed on our own.

In [ ]:
speeds = coaster_df['Top_Speed']

Mean

In [ ]:
mean = speeds.mean()
print("The mean of top speed is", mean)

Median

In [ ]:
median = speeds.median()
print("The median top speed is", median)

Mode

In [ ]:
mode = speeds.mode()
print("The most common top speed is", mode)

Notice that ```mode``` returned two numbers: 0 and 50. That means that both are equally common. The ```dtype``` tells you that the mode is a number (```float64```).

Standard Deviation

In [ ]:
stdev = speeds.std()
print("The standard top speed of depths is", stdev)

## Visualize

Remember that matplotlib has some helpful functions to visualize these statistics. For example, ```.boxplot```

In [ ]:
import matplotlib.pyplot as plt

plt.boxplot(speeds)
plt.show()

I could add to the boxplot function to visualize without the outliers too.

In [ ]:
plt.boxplot(speeds, showfliers=False)
plt.show()

Another common method for visualizing descriptive statistics is a histogram.

In [ ]:
plt.hist(speeds)
plt.show()

I could visualize the mean within the histogram using ```axvline(***.mean())```

In [ ]:
plt.hist(coaster_df['Top_Speed'])
plt.axvline(speeds.mean(), color = "red") #make it red so we can see it
plt.show()

## Final Code

If I were to put all of my code together, I would do it like this.



In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

#Read and clean in data
coaster_df = pd.read_csv('/content/roller_coaster.csv')
coaster_df.dropna(inplace = True, columns = ['Top_Speed'])
print(coaster_df.describe())

#calculate statistics
mean = speeds.mean()
median = speeds.median()
mode = speeds.mode()
stdev = speeds.std()

#print results
print("The mean top speed is", mean)
print("The median top speed is", median)
print("The most common top speed is", mode)
print("The standard top speed of depths is", stdev)

#Create visualizations
plt.boxplot(speeds)
plt.show()

plt.hist(coaster_df['Top_Speed'])
plt.axvline(speeds.mean(), color = "red")
plt.show()



## Your Turn

Upload a dataset of your choice. Clean the data set,  Calculate the mean, median, mode, and standard deviation of one of the columns.

Then, visualize the statistics by creating a boxplot and a histogram.

In [ ]:
url = 'https://drive.google.com/file/d/10tbIWJ236s0JnKFC5JyeAN_Igd_Yshlp/view?usp=sharing'
output_path = 'my_data.csv'
gdown.download(url, output_path, quiet=False, fuzzy=True)
my_df = pd.read_csv('/content/my_data.csv')

Downloading...
From: https://drive.google.com/uc?id=10tbIWJ236s0JnKFC5JyeAN_Igd_Yshlp
To: /content/my_data.csv
100%|██████████| 1.43k/1.43k [00:00<00:00, 4.12MB/s]
